### ISSUE TITLE
[RFC] Splitting benchmarks into datasets

### ISSUE THREAD

**Issue by ldionne** [ldionne/metabench#15]
Here's an idea I've been thinking about; I also started implementing it, but it's very rough right now. The idea is to allow benchmarks to be described in `CMakeLists.txt` files. For example, you would have a directory structure

```
path/
    to/
        benchmark/
            CMakeLists.txt
            file1.cpp.erb
            ...
            fileN.cpp.erb
```

The `.cpp.erb` files would be just the same as the current `.cpp` files, i.e. they would contain the code to be benchmarked in the form of ERB templates. The `CMakeLists.txt` file would look somewhat like this:

```cmake
add_dataset(file1 "(0..100).step(5)") # <- the name of the file and the range to benchmark
add_dataset(fileN "....)
add_benchmark(name_of_the_benchmark file1 ... fileN)
```

The `add_benchmark` call would add a target that runs the benchmarks and gathers all the datasets (a bit like we do right now), but the dependency between a benchmark and the `.cpp` files it "contains" would be stated explicitly in CMake. Ok, but what are the advantages to doing this?

1. We could allow setting CMake properties to each dataset (such as include directories, flags, etc...) in a finer-grained manner, while still being portable if we run the benchmarks through CMake (as suggested in #14).
2. Users wouldn't have to worry about how the charts are specified. They would only have to call the CMake commands and we would generate the proper JSON files for them.
3. We could provide different kinds of benchmarks, for example compile-time and compiler RAM usage. This could be achieved by adding options to `add_benchmark` (like `add_benchmark(name [COMPILE_TIME|RAM] file1 ... fileN)`). For each kind of benchmark, we would have a simple template that sets reasonable defaults for chart visualization (for compile-time the yAxis is seconds, and we can set the titles of the axis ourselves).
4. We could exclude datasets from a benchmark based on CMake conditions naturally, without resorting to `cmake_bool`.

The only downside I see to this approach so far is that it's slightly less flexible, in the sense that being able to write arbitrary Ruby code in the `chart.json` file is pretty damn flexible. If we take over the control of `chart.json` and generate it ourselves from CMake, then we're taking away some flexibility, for better or worse.

Any thoughts?

**Comment by brunocodutra** [comment]
I guess all the flexibility one needs may be expressed directly in the .cpp files.

Now, to retain the ability of configuring the chart's properties, we could mix both approaches and have CMake insert the `data:` property to the JSON provided in the chart.json file, which would then exclusivelly be used to set the chart's properties. If no chart.json were provided, then CMake assumes an empty object (thus relying on the default properties hardcoded on the index.html template)

**Comment by brunocodutra** [comment]
Alright, I made a basic implementation of this idea. It still needs a fair deal of polishing, specially the ruby part, but it does work already.

First one adds datasets using `metabench_add_dataset`, which takes an identifier, a source file, a ruby range as a string and the environment also as a string

```
metabench_add_dataset(std_decay decay/std_decay.cpp.erb "(0...500).step(100)" "{}")
metabench_add_dataset(my_decay decay/my_decay.cpp.erb "(0...500).step(100)" "{var: foo}")
```

When built, a dataset produces a `.json` file containing its benchmark results for all n within `_metabench/`. Something like this:

```
"data": [{                                                   
  key: "std_decay",                                                  
  values: [[0, 0.023423962000379106], [100, 0.08739784000044892], [200, 0.14871006500015937], [300, 0.19648170299933554], [400, 0.26728218799962633]]                                                           
}, {                                                   
  key: "my_decay",                                                  
  values: [[0, 0.022123990999716625], [100, 0.053515164000600635], [200, 0.08566520899967145], [300, 0.11931800599995768], [400, 0.1476865140002701]]                                                           
}]

```
To add benchmarks, one then uses `metabench_add_benchmark`, which takes an identifier and a list of datasets

```
metabench_add_benchmark(decay std_decay my_decay)
```
When built, a benchmark reads the `.json` files produced by each dataset and outputs the usual `.html` file and also another `.json` file containing an array of the individual benchmarks.

Now to make it possible for the user to configure the chart, I'm thinking `metabench_add_benchmark` could have an optional parameter intended for the user to supply a `.json` file containing the structure of the current `chart.json` except for the `data:` attribute, which would be automatically inserted.

So, what do you think?

**Comment by ldionne** [comment]
I think that looks good. Can you submit a PR so we can discuss?

**Comment by brunocodutra** [comment]
Sure, in a few minutes.

**Comment by ldionne** [comment]
This is resolved by #21.
